iree/tools/iree-e2e-matmul-test.c

```diff
@@ -475,106 +681,364 @@ static iree_status_t check_matmul_impl(
-static iree_status_t allocate_buffer_like(iree_hal_allocator_t* hal_allocator,
-                                          iree_hal_buffer_view_t* src,
-                                          iree_hal_buffer_view_t** dst) {
-  return iree_hal_buffer_view_allocate_buffer(
-      hal_allocator, iree_hal_buffer_view_shape_dims(src),
-      iree_hal_buffer_view_shape_rank(src),
-      iree_hal_buffer_view_element_type(src),
-      iree_hal_buffer_view_encoding_type(src),
-      (iree_hal_buffer_params_t){
-          .type = IREE_HAL_MEMORY_TYPE_HOST_LOCAL,
-          .usage =
-              IREE_HAL_BUFFER_USAGE_TRANSFER | IREE_HAL_BUFFER_USAGE_MAPPING,
-      },
-      iree_const_byte_span_empty(), dst);
+/*****************************************************************************
+ *
+ * Part 4:
+ *
+ * Core matmul test logic.
+ *
+ * The entry point into this part is |replay_event_call_matmul|, which is the
+ * handler for each matmul testcase in the trace. Other functions are only
+ * helpers for it.
+ *
+ * |replay_event_call_matmul| calls |do_matmul_and_check_results| to actually
+ * perform a matmul. In normal cases, each |replay_event_call_matmul| performs
+ * one call to |do_matmul_and_check_results|, but when that generates an error,
+ * it will make additional calls to |do_matmul_and_check_results| to evaluate
+ * variants of the failed testcase to generate a more helpful log.
+ *
+ * The |matrix_mask_t| stuff is only used to generate these variants of failed
+ * testcases.
+ *
+ *****************************************************************************/
+
+// Enumerates ways that we may mask matrices in list of matrix inputs to matmul
+// testcases.
+typedef enum {
+  MATRIX_MASK_NONE,      // no-op: leave the existing matrix unchanged.
+  MATRIX_MASK_ZERO,      // overwrite the matrix with zeros.
+  MATRIX_MASK_IDENTITY,  // overwrite with (general rectangular) identity matrix
+} matrix_mask_t;
+
+// Allocates device-local |dst| and initializes it as an identity-matrix shaped
+// like |src|.
+static iree_status_t make_device_identity_matrix_like(
+    iree_hal_device_t* device, iree_hal_allocator_t* hal_allocator,
+    iree_hal_buffer_view_t* src, iree_hal_buffer_view_t** dst) {
+  iree_hal_element_type_t elem_type = iree_hal_buffer_view_element_type(src);
+  iree_host_size_t elem_byte_count =
+      iree_hal_element_dense_byte_count(elem_type);
+  iree_hal_dim_t dims[2] = {0};
+  IREE_RETURN_IF_ERROR(get_matrix_shape(src, dims));
+  int rows = dims[0];
+  int cols = dims[1];
+  iree_hal_buffer_view_t* host_dst = NULL;
+  IREE_RETURN_IF_ERROR(
+      allocate_host_buffer_view_like(hal_allocator, src, &host_dst));
+  void* host_dst_data = NULL;
+  IREE_CHECK_OK(map_host_local_row_major_data(
+      host_dst, IREE_HAL_MEMORY_ACCESS_WRITE, &host_dst_data));
+  // Write 1 to matrix elements on the main diagonal.
+  int diagonal_size = iree_min(rows, cols);
+  for (int i = 0; i < diagonal_size; ++i) {
+    write_int_element(
```

Comment:
All of this can be simplified using the `iree_hal_buffer_view_generate_buffer` method - that does the allocation, mapping (if needed), and transfer. trace_replay.c uses this to parse/generate buffers like this today (that's why `iree_trace_replay_generate_identity_matrix` was so simple in there) and image_util.c has an example that loads an image into a buffer. You basically pass in your buffer params (which can be just DEVICE_LOCAL) and a callback that gets made with a host-accessible void* you can populate with whatever you want (like this loop).

Reply:
Happy to use `iree_hal_buffer_view_generate_buffer` ! --- but FWIW, here is the current `iree_trace_replay_generate_identity_matrix`, it's not using it: https://github.com/google/iree/blob/08e45ea3e1856967dabc845e34393dae48dfce3f/iree/tools/utils/trace_replay.c#L550-L567

Reply:
(At first I didn't understand as `iree_trace_replay_generate_identity_matrix` doesn't use `iree_hal_buffer_view_generate_buffer` but now I get it, it's the other way around).

Reply:
Done.